src/core/server/saved_objects/service/lib/repository.test.js

```diff
@@ -356,2644 +240,2817 @@ describe('SavedObjectsRepository', () => {
+        await expect(
+          savedObjectsRepository.addNamespaces(type, id, namespaces, options)
+        ).rejects.toThrowError(createGenericNotFoundError(type, id));
+      };
+      const expectBadRequestError = async (type, id, namespaces, message) => {
+        await expect(
+          savedObjectsRepository.addNamespaces(type, id, namespaces)
+        ).rejects.toThrowError(createBadRequestError(message));
+      };
+
+      it(`throws when type is invalid`, async () => {
+        await expectNotFoundError('unknownType', id, [newNs1, newNs2]);
+        expect(callAdminCluster).not.toHaveBeenCalled();
       });
 
-      expect(callAdminCluster).toHaveBeenCalledTimes(1);
-      expect(callAdminCluster).toHaveBeenCalledWith(
-        'index',
-        expect.objectContaining({
-          index: '.kibana-test',
-        })
-      );
-    });
+      it(`throws when type is hidden`, async () => {
+        await expectNotFoundError(HIDDEN_TYPE, id, [newNs1, newNs2]);
+        expect(callAdminCluster).not.toHaveBeenCalled();
+      });
 
-    it('should use custom index', async () => {
-      await savedObjectsRepository.create('baz', {
-        id: 'logstash-*',
-        title: 'Logstash',
+      it(`throws when type is not namespace-agnostic`, async () => {
+        const test = async type => {
+          const message = `${type} doesn't support namespaces`;
+          await expectBadRequestError(type, id, [newNs1, newNs2], message);
+          expect(callAdminCluster).not.toHaveBeenCalled();
+        };
+        await test('index-pattern');
+        await test(NAMESPACE_AGNOSTIC_TYPE);
       });
 
-      expect(callAdminCluster).toHaveBeenCalledTimes(1);
-      expect(callAdminCluster).toHaveBeenCalledWith(
-        'index',
-        expect.objectContaining({
-          index: 'beats',
-        })
-      );
-    });
+      it(`throws when namespaces is not an array of strings`, async () => {
+        const test = async namespaces => {
+          const message = 'namespaces must be a non-empty array of strings';
+          await expectBadRequestError(type, id, namespaces, message);
+          expect(callAdminCluster).not.toHaveBeenCalled();
+        };
+        await test('string');
+        await test(123);
+        await test(true);
+        await test(null);
+        await test(undefined);
+        await test([]);
+      });
 
-    it('migrates the doc', async () => {
-      migrator.migrateDocument = doc => {
-        doc.attributes.title = doc.attributes.title + '!!';
-        doc.migrationVersion = { foo: '2.3.4' };
-        doc.references = [{ name: 'search_0', type: 'search', id: '123' }];
-        return doc;
-      };
+      it(`throws when ES is unable to find the document during get`, async () => {
+        callAdminCluster.mockResolvedValue({ found: false }); // this._callCluster('get', ...)
+        await expectNotFoundError(type, id, [newNs1, newNs2]);
+        expect(callAdminCluster).toHaveBeenCalledTimes(1);
```

Comment:
I am hesitant to check for _all_ cluster call args -- I specifically divided the 'cluster call' test cases into their own subset of cases for each function, so we can divide concerns with each test case.

However, looking at this, I agree that we should at least be checking the cluster call _actions_ (`get`, `update`, etc.) for the error test cases, since those can vary depending on the test case.

Let me go ahead and make that change, then you can take another look and see if that makes sense to you as well. If not, I'll revisit this again.